applications/FreeSurfaceApplication/custom_utilities/edge_data_c2c.h

```diff
@@ -0,0 +1,1374 @@
+                dest[comp] = orig[comp];
+        }
+    }
+
+    void AssignVectorToVector(const ValuesVectorType& origin,
+                              ValuesVectorType& destination
+                             )
+    {
+        int loop_size = origin.size();
+        #pragma omp parallel for
+        for (int i_node = 0; i_node < loop_size; i_node++)
+        {
+            destination[i_node] = origin[i_node];
+        }
+    }
+
+
+
+private:
+    //number of edges
+    unsigned int mNumberEdges;
+
+    //CSR data vector for storage of the G, L and consistent M components of edge ij
+    EdgesVectorType mNonzeroEdgeValues;
+
+    //vector to store column indices of nonzero matrix elements for each row
+    IndicesVectorType mColumnIndex;
+
+    //index vector to access the start of matrix row i in the column vector
+    IndicesVectorType mRowStartIndex;
+
+    //inverse of the mass matrix ... for parallel calculation each subdomain should contain this correctly calculated (including contributions of the neighbours)
+    ValuesVectorType mInvertedMassMatrix;
+
+    //minimum height around one node
+    ValuesVectorType mHmin;
+
+    //lumped mass matrix (separately stored due to lack of diagonal elements of the consistent mass matrix)
+    ValuesVectorType mLumpedMassMatrix;
+    //diagonal of the gradient matrix (separately stored due to special calculations)
+    CalcVectorType mDiagGradientMatrix;
+
+    //*******************************************
+    //functions to set up elemental mass matrices
+
+    void CalculateMassMatrix(boost::numeric::ublas::bounded_matrix<double, 3, 3 > & mass_consistent, double volume)
+    {
+        for (unsigned int i_node = 0; i_node <= TDim; i_node++)
+        {
+            //diagonal terms
+            mass_consistent(i_node, i_node) = 0.16666666666666666667 * volume; //1/6
+            //non-diagonal terms
+            double temp = 0.08333333333333333333 * volume; // 1/12
+            for (unsigned int j_neighbour = i_node + 1; j_neighbour <= TDim; j_neighbour++)
+            {
+                //taking advantage of symmetry
+                mass_consistent(i_node, j_neighbour) = temp;
+                mass_consistent(j_neighbour, i_node) = temp;
+            }
+        }
+    }
+
+    void CalculateMassMatrix(boost::numeric::ublas::bounded_matrix<double, 4, 4 > & mass_consistent, double volume)
+    {
+        for (unsigned int i_node = 0; i_node <= TDim; i_node++)
+        {
+            //diagonal terms
+            mass_consistent(i_node, i_node) = 0.1 * volume;
+            //non-diagonal terms
+            double temp = 0.05 * volume;
+            for (unsigned int j_neighbour = i_node + 1; j_neighbour <= TDim; j_neighbour++)
+            {
+                //taking advantage of symmetry
+                mass_consistent(i_node, j_neighbour) = temp;
+                mass_consistent(j_neighbour, i_node) = temp;
+            }
+        }
+    }
+
+};
+
+} //namespace Kratos
+
```

Comment:
Indeed most of the code is the same, can your create a base class and override the specific methods you need?